File: src/main/java/com/und/security/RestAuthenticationProvider.java
package com.und.security;

import com.und.security.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
import org.springframework.stereotype.Component;

/**
 * Created by shiv on 14/08/17.
 */
@Component
public class RestAuthenticationProvider implements AuthenticationProvider {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private RestTokenUtil restTokenUtil;

    @Override
    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
        RestAuthenticationToken token = (RestAuthenticationToken) authentication;
        EventUser user = RestUserFactory.create(userRepository.findByUsername(token.getName()));

        if(user != null || token.getKey().equalsIgnoreCase(user.getKey())) {
            if (restTokenUtil.validateToken(user.getKey(), user)) {
                return new RestAuthenticationToken(user, user.getPassword(), user.getAuthorities(), token.getKey());
            }
        }


        throw new BadCredentialsException("The credentials are invalid");

    }

    @Override
    public boolean supports(Class<?> authentication) {
        return RestAuthenticationToken.class.equals(authentication);
    }

}

File: src/main/kotlin/com/und/security/RestAuthenticationProvider.kt
package com.und.security

import com.und.security.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.authentication.AuthenticationProvider
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.Authentication
import org.springframework.security.core.AuthenticationException
import org.springframework.stereotype.Component

/**
 * Created by shiv on 14/08/17.
 */
@Component
class RestAuthenticationProvider : AuthenticationProvider {

    @Autowired
    private val userRepository: UserRepository? = null

    @Autowired
    private val restTokenUtil: RestTokenUtil? = null

    @Throws(AuthenticationException::class)
    override fun authenticate(authentication: Authentication): Authentication {
        val token = authentication as RestAuthenticationToken
        val user = RestUserFactory.create(userRepository!!.findByUsername(token.getName())!!)

        if (user != null || token.key.equals(user.key)) {
            if (restTokenUtil!!.validateToken(user.key!!, user)) {
                return RestAuthenticationToken(user, user.password!!, user.authorities!!, token.key)
            }
        }


        throw BadCredentialsException("The credentials are invalid")

    }

    override fun supports(authentication: Class<*>): Boolean {
        return RestAuthenticationToken::class.java == authentication
    }

}